src/main/java/org/reaktivity/reaktor/internal/agent/ElektronAgent.java

```diff
@@ -910,20 +910,23 @@ private void handleDefaultReadInitial(
             break;
         case DataFW.TYPE_ID:
             final DataFW data = dataRO.wrap(buffer, index, index + length);
-            handleDefaultData(data);
+            final long traceId = data.traceId();
+            final long budgetId = data.budgetId();
+            final int reserved = data.reserved();
+
+            doSystemWindowIfNecessary(traceId, budgetId, reserved);
             break;
         }
     }
 
-    private void handleDefaultData(
-        DataFW data)
+    private void doSystemWindowIfNecessary(
+        long traceId,
+        long budgetId,
+        int reserved)
     {
-        final long traceId = data.traceId();
-        final long budgetId = data.budgetId();
-        final int reserved = data.reserved();
-
         if (budgetId != 0L)
         {
+            assert reserved > 0;
```

Comment:
what if `DATA` payload is `null` and `reserved` is `0`, but `budgetId` is non-zero?
the reserved check should be part of the condition, and it also acts as an optimization since crediting `0` is a no-op anyway